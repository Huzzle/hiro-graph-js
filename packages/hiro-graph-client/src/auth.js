const filterUndef = obj =>
    Object.keys(obj).reduce((ret, k) => {
        if (obj[k] !== undefined) {
            ret[k] = obj[k];
        }
        return ret;
    }, {});

export const auth = {
    organizationTeams: (fetch, options, id) =>
        fetch(`/api/6.1/iam/organization/${id}/teams`, options),
    updateMeProfile: (fetch, options, data) => {
        const payload = filterUndef(data);

        options.method = "POST";
        options.body = JSON.stringify(payload);
        options.headers["Content-Type"] = "application/json";

        return fetch(`/api/6.1/iam/me/profile`, options);
    },
    mePassword: (fetch, options, oldPassword, newPassword) => {
        options.method = "POST";
        options.body = JSON.stringify({
            oldPassword,
            newPassword
        });
        options.headers["Content-Type"] = "application/json";

        return fetch(`/api/6.1/iam/me/password`, options);
    },
    updateAccountProfile: (fetch, options, id, data) => {
        const payload = filterUndef(data);

        options.method = "POST";
        options.body = JSON.stringify(payload);
        options.headers["Content-Type"] = "application/json";

        return fetch(`/api/6.1/iam/accounts/profile/${id}`, options);
    },
    getAvatar: (fetch, options, id) =>
        fetch(`/api/6.1/iam/accounts/${id}/avatar`, { ...options, raw: true })
};
