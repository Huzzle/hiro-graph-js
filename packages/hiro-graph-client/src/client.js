/**
 *  A GraphIT Client
 *
 *  It will use Websockets where possible and fall back to HTTP where WebSockets not supported.
 *  Event streams require WebSockets so those will not work witout websocket support.
 *  The connection requires a `Token`.
 */
import WebsocketTransport, {
    webSocketsAvailable
} from "./transport-websocket-pool";
import HttpTransport from "./transport-http";
//import EventTransport from "./eventstream";
import {
    isUnauthorized,
    isTransactionFail,
    isConflict,
    isNotFound,
    connectionClosedBeforeSend
} from "./errors";

const passthru = fn => [
    r => (fn(), r),
    e => {
        fn();
        throw e;
    }
];

//nb this isn't a deep clone, just a top level deref.
//we only deal with primitives, arrays and plain objects.
const dereference = obj => {
    if (typeof obj !== "object" || obj === null) {
        return obj;
    }
    if (Array.isArray(obj)) {
        return obj.slice();
    }
    return Object.assign({}, obj);
};

export default class Client {
    constructor({ endpoint, token }, transportOptions = {}) {
        this.endpoint = endpoint;

        //we hold on to the token for ease of access/manual invalidation
        this.token = token;

        //create the transports.
        //the http is for http only endpoint, e.g. the /_ki/* and /_variable/* servlets
        this.http = new HttpTransport(endpoint);

        // This is the main transport for most of the api.
        // Use Websocket if available (except if specifically disabled)
        // fallback to http, or use given transport
        // duck type the final parameter
        if (typeof transportOptions.request === "function") {
            this.transport = transportOptions;
        } else if (webSocketsAvailable && !transportOptions.forceHTTP) {
            //All a transport needs to implement is "request"
            this.transport = new WebsocketTransport(endpoint, transportOptions);
        } else {
            if (!webSocketsAvailable) {
                console.warn(
                    "WebSockets not available, falling back to HTTP transport"
                );
            }
            this.transport = this.http;
        }

        // This is where we keep our event streams.
        // NB EventStream is not implemented yet
        this.events = [];

        // Bind our fetch for extension servlets.
        this.fetch = (...args) => this.http.fetch(this.token, ...args);

        this._dedup = Object.create(null);

        //keep this so we can duplicate them
        this._servlets = [];
    }

    debugRequests(turnOnDebugging) {
        this._debug_requests = turnOnDebugging === true;
    }

    /**
     *  Make a clone of this connection, but with new
     */
    cloneWithNewToken(newToken) {
        const cloned = new Client(
            {
                endpoint: this.endpoint,
                token: newToken
            },
            this.transport
        );
        this._servlets.forEach(([name, methods]) =>
            cloned.addServlet(name, methods)
        );
        return cloned;
    }

    /**
     *  Make a generic request. The transport will handle the details, we handle the retry.
     *  We use `this.transport` which is websocket if available.
     */
    request(
        { type, headers = {}, body = {} } = {},
        reqOptions = {},
        retries = 1
    ) {
        return this.transport
            .request(
                this.token,
                { type, headers, body },
                Object.assign({ debug: this._debug_requests }, reqOptions)
            )
            .catch(err => {
                //these are the special cases.
                //regular errors end up with code === undefined, so not retryable.
                switch (true) {
                    case err === connectionClosedBeforeSend:
                        //this means unconditionally retry
                        return this.request(
                            { type, headers, body },
                            reqOptions,
                            retries
                        );
                    case isUnauthorized(err): //unauthorized (which means unauthenticated) invalidate TOKEN.
                        this.token.invalidate();
                        err.isRetryable = true;
                        break;
                    case isTransactionFail(err): //error persisting transaction. retryable.
                        err.isRetryable = true;
                        break;
                    //there are other known errors, e.g. 403, 400, etc... but they are not retryable.
                    default:
                        if ("isRetryable" in err === false) {
                            err.isRetryable = false;
                        }
                        break;
                }
                //a chance to retry
                if (err.isRetryable && retries > 0) {
                    return this.request(
                        { type, headers, body },
                        reqOptions,
                        retries - 1
                    );
                }
                throw err;
            });
    }

    // Deduplicates a request, but the request must be an Array returning one.
    dedupedRequest(
        { type, headers = {}, body = {} } = {},
        reqOptions = {},
        retries = 1
    ) {
        // isDebug is either set in requestion options, or if not, it is inherited from `this`
        const isDebug = "debug" in reqOptions
            ? reqOptions.debug
            : this._debug_requests;
        try {
            const requestKey = JSON.stringify({ type, headers, body });
            if (isDebug) {
                console.info(
                    "deduped request, key:",
                    requestKey,
                    "isDupe?",
                    requestKey in this._dedup
                );
            }
            if (requestKey in this._dedup) {
                this._dedup[requestKey]._calls++;
                return this._dedup[requestKey].then(dereference);
            }
            const cleanUp = passthru(() => {
                if (isDebug) {
                    const calls = this._dedup[requestKey]._calls;
                    if (calls > 1) {
                        console.info("Depuped Request (%dx)", calls, {
                            type,
                            headers,
                            body
                        });
                    }
                }
                delete this._dedup[requestKey];
            });
            const promise = (this._dedup[requestKey] = this.request(
                { type, headers, body },
                reqOptions,
                retries
            )
                .then(...cleanUp)
                .then(res => (promise._calls > 1 ? dereference(res) : res)));
            promise._calls = 1;
            return promise;
        } catch (e) {
            console.error("dedup error", e.stack);
            return Promise.reject(e);
        }
    }

    getToken() {
        return this.token;
    }

    /**
     *  Core GraphIT REST API methods
     *  Prefer websocket transport
     *  The servlets for e.g. /_ki/validate are added using mixins
     */

    /**
     *  GraphIT Server Info
     */
    info(reqOptions = {}) {
        return this.request({ type: "info" }, reqOptions);
    }

    /**
     *  Get a single item by ID
     */
    get(id = "_me", reqOptions = {}) {
        if (id === "_me") {
            return this.me(reqOptions);
        }
        return this.dedupedRequest(
            { type: "get", headers: { "ogit/_id": id } },
            reqOptions
        );
    }

    /**
     *  Get the node for the owner of this token
     */
    me(reqOptions = {}) {
        return this.dedupedRequest({ type: "me" }, reqOptions);
    }

    /**
     *  Create a new node
     */
    create(type, data = {}, reqOptions = {}) {
        const headers = { "ogit/_type": type };
        if (reqOptions.waitForIndex) {
            headers.waitForIndex = "true";
        }
        return this.request(
            { type: "create", headers, body: data },
            reqOptions
        );
    }

    /**
     *  Update a Node
     */
    update(id, data = {}, reqOptions = {}) {
        const headers = { "ogit/_id": id };
        if (reqOptions.waitForIndex) {
            headers.waitForIndex = "true";
        }
        return this.request(
            { type: "update", headers, body: data },
            reqOptions
        );
    }

    /**
     *  Replace a Node, optionally `upsert`
     */
    replace(id, data = {}, reqOptions = {}) {
        const { createIfNotExists = false, waitForIndex = false } = reqOptions;
        //createIfNotExists should contain the "ogit/_type" to create if the node doesn't exist,
        //and nothing otherwise
        const headers = { "ogit/_id": id };
        if (createIfNotExists) {
            Object.assign(headers, {
                createIfNotExists: "true",
                "ogit/_type": createIfNotExists
            });
        }
        if (waitForIndex) {
            headers.waitForIndex = "true";
        }
        return this.request(
            { type: "replace", headers, body: data },
            reqOptions
        );
    }

    /**
     *  Delete a node/edge
     *
     *  We don't handle the 404/409 here. or should we?
     */
    delete(id, reqOptions = {}) {
        const { waitForIndex = false } = reqOptions;
        const headers = { "ogit/_id": id };
        if (waitForIndex) {
            headers.waitForIndex = "true";
        }
        return this.request({ type: "delete", headers }, reqOptions);
    }

    /**
     *  This is a vertices query
     */
    lucene(
        query = "",
        {
            limit = 50,
            offset = 0,
            order = false,
            fields = [],
            count = false,
            ...placeholders
        } = {},
        reqOptions = {}
    ) {
        const body = {
            query,
            limit,
            offset,
            count,
            ...placeholders
        };
        if (order) {
            body.order = "" + order; // this implicitly does array.join(",") for arrays. (and works with strings...)
        }
        if (fields.length > 0) {
            body.fields = fields.join(",");
        }
        return this.dedupedRequest(
            {
                type: "query",
                headers: { type: "vertices" },
                body
            },
            reqOptions
        );
    }

    ids(list, reqOptions = {}) {
        return this.dedupedRequest(
            {
                type: "query",
                headers: { type: "ids" },
                body: { query: list.join(",") } // yes, it has to be a comma-seperated string
            },
            reqOptions
        );
    }

    /**
     *  This is a gremlin query
     */
    gremlin(root, query, reqOptions = {}) {
        if (this._debug_requests) {
            console.log(`GRAPHIT#gremlin(${root}, ${query})`);
        }
        return this.dedupedRequest(
            {
                type: "query",
                headers: { type: "gremlin" },
                body: { root, query }
            },
            reqOptions
        );
    }

    /**
     *  Connect two Nodes with an edge of `type`
     */
    connect(type, inId, outId, reqOptions = {}) {
        return this.request(
            {
                type: "connect",
                headers: { "ogit/_type": type },
                body: { in: inId, out: outId }
            },
            reqOptions
        ).then(
            () => {}, //return nothing.
            err => {
                //Conflict is OK here, just means that the edge was already connected.
                if (isConflict(err)) {
                    return; //return nothing.
                }
                //real error.
                throw err;
            }
        );
    }

    /**
     *  Disconnect two nodes, convenience for delete, generates the edge id for you.
     */
    disconnect(type, inId, outId, reqOptions = {}) {
        return this.delete(`${inId}$$${type}$$${outId}`, reqOptions).then(
            () => {}, //return nothing.
            err => {
                //Not Found or Conflict is OK here, just means that the edge was already deleted/didn't ever exist
                if (isNotFound(err) || isConflict(err)) {
                    return; //return nothing.
                }
                //real error.
                throw err;
            }
        );
    }

    /**
     *  Custom servlet endpoints
     *  always use HTTP transport, which knows how to handle these types.
     *
     *  The servletMethods are passed:
     *  - `fetch` a token/endpoint aware HTTP fetch API.
     *  - `options` the default fetch options you can override
     *  - `...args` the rest of the args passed in by the user for that method
     */
    addServlet(prefix, servletMethods) {
        if (!prefix) {
            throw new Error("[GRAPH] Must give prefix for servlet");
        }
        if (prefix in this) {
            throw new Error(
                "[GRAPH] Sevlet Extensions must have unique prefixes. Attempted to re-add `" +
                    prefix +
                    "`"
            );
        }
        //create namespace.
        this[prefix] = {};
        //add servlet methods
        Object.keys(servletMethods).reduce((acc, method) => {
            acc[method] = (...args) => {
                const call = servletMethods[method];
                return call(
                    this.fetch,
                    this.http.defaultOptions(),
                    ...args
                ).catch(err => {
                    //these are the special cases.
                    //regular errors end up with code === undefined, so not retryable.
                    switch (true) {
                        case isUnauthorized(err): //unauthorized (which means unauthenticated) invalidate TOKEN.
                            this.token.invalidate();
                            err.isRetryable = true;
                            break;
                        case isTransactionFail(err): //error persisting transaction. retryable.
                            err.isRetryable = true;
                            break;
                        //there are other known errors, e.g. 403, 400, etc... but they are not retryable.
                        default:
                            if ("isRetryable" in err === false) {
                                err.isRetryable = false;
                            }
                            break;
                    }
                    //a chance to retry - only once.
                    if (err.isRetryable) {
                        return servletMethods[method](
                            this.fetch,
                            this.http.defaultOptions(),
                            ...args
                        );
                    }
                    throw err;
                });
            };
            return acc;
        }, this[prefix]);
        //chainable? might be useful.
        return this;
    }
}
