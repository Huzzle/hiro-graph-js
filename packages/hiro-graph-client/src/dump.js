import { inspect } from "util";

export default function dump(obj) {
    return inspect(obj, { depth: null, colors: true });
}
