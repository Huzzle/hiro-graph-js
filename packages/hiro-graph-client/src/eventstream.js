/**
 *  Each instance of events stream subscribes to the /_events/ endpoint of GraphIT
 *  Filters can be applied to each stream and events are emitted.
 *  The GraphIT connection does not keep hold of these eventstreams, the consumer does.
 *  The streams are clever enough to keep track of their position and reconnect (with current
 *  filters) on the event of connection failure. Should the token expire and connection fail,
 *  if it is renewed by any other part of the app that shares the `Token` object, it will reconnect
 *  and continue to emit events.
 */
import { ensureWebSocketsAvailable } from "./transport-websocket";

//unique ids.
let _id = 0;
const nextId = () => {
    const n = _id++;
    if (_id > 1e9) {
        //this has got big. let's reset.
        _id = 0;
    }
    return "" + n; //needs to be a string.
};

const cloneEvent = evt => {
    //the only "deep" prop is the "body", and it's just one level deep.
    return Object.assign({}, evt, { body: Object.assign({}, evt.body) });
};

export default class EventStream {
    constructor({ endpoint, token }) {
        ensureWebSocketsAvailable();
        this.token = token;
        this.endpoint = endpoint
            .replace(/^http/, "ws") //replace http(s) with ws(s)
            .replace(/\/?$/, "/_events/?_TOKEN="); // replace possible trailing slash with api endpoint
        this.callbacks = [];
        this.filters = [];
        //we don't connect immediately.
    }

    connected() {
        return this.__connectionPromise;
    }

    //stop events.
    stop() {
        this.stopped = true;
        if (!this.connected()) {
            return;
        }
        return this.connect().then(ws => {
            ws.close();
        });
    }

    //start events
    start() {
        this.stopped = false;
        if (this.connected()) {
            return;
        }
        //don't actually return the connection.
        return this.connect().then(() => {});
    }

    //Private function used to broadcast events.
    emit(event) {
        //return as a promise in case we want to wait for the callbacks.
        return Promise.all(this.callbacks.map(fn => fn(cloneEvent(event))));
    }

    //Start a subscription to events. returns a function used to remove the subscription
    subscribe(callback) {
        if (typeof callback !== "function") {
            throw new TypeError("callback must be a function");
        }
        this.callbacks.push(callback);
        //return the unsubscribe function.
        return () =>
            (this.callbacks = this.callbacks.filter(fn => fn !== callback));
    }

    //Register a filter to the stream.
    //returns the unregister function.
    filter(type, content) {
        const id = nextId();
        const filter = {
            "filter-id": id,
            "filter-type": type,
            "filter-content": content
        };
        this.filters.push(filter);
        const unfilter = () => {
            this.filters = this.filters.filter(f => f["filter-id"] !== id);
            return (
                this.connected() &&
                this.connect().then(ws =>
                    ws.send(
                        JSON.stringify({
                            type: "unregister",
                            args: { "filter-id": id }
                        })
                    )
                )
            );
        };

        if (this.connected()) {
            return this.connect()
                .send({
                    type: "register",
                    args: filter
                })
                .then(() => unfilter);
        }
        return Promise.resolve(unfilter);
    }

    //clear all filters.
    clear() {
        this.filters = [];
        return this.connect().then(ws =>
            ws.send(
                JSON.stringify({
                    type: "clear",
                    args: {}
                })
            )
        );
    }

    connect() {
        if (!this.__connectionPromise) {
            this.__connectionPromise = this.token
                .get()
                .then(token => this.createWebSocket(token));
        }
        return this.__connectionPromise;
    }

    /**
     *  Creates a WebSocket connection, with the initial token.
     *
     *  Similiar logic to the websocket-transport
     */
    createWebSocket(intialToken) {
        //we do all this in a promise, so the result is shared between
        //all requests that come in during connection.
        return new Promise((resolve, reject) => {
            const ws = new WebSocket(this.endpoint + intialToken);

            //this keeps track of whether we have resolved yet.
            let hasResolved = false;

            //when the connection closes we need to tidy up.
            ws.onclose = () => {
                //most importantly, remove this promise. so a new one can be created.
                this.__connectionPromise = null;

                //if we haven't resolved yet, then this is an *immediate* close by GraphIT.
                //therefore we probably have an invalid token, and we should definitely reject.
                if (!hasResolved) {
                    hasResolved = true;
                    this.token.invalidate();
                    reject();
                }
                //We want to reconnect unless this.stopped
                if (!this.stopped) {
                    this.connect();
                }
            };

            //on connection error, we almost certainly close after this anyway, so just
            //log the error
            ws.onerror = err => {
                console.error("[EVT] WebSocket Connection Error", err);
            };

            //how to handle inbound messages, find the inflight it matches and return
            ws.onmessage = msg => {
                //we can only handle string messages
                if (typeof msg.data !== "string") {
                    //we don't handle messages like this.
                    console.warn("[EVT] unknown websocket message type", msg);
                    return;
                }
                //try and parse as JSON
                let object;
                try {
                    object = JSON.parse(msg.data);
                } catch (e) {
                    console.warn(
                        "[EVT] invalid JSON in websocket message",
                        msg.data
                    );
                    return;
                }
                //this event should have properties:
                //
                //  id: the id of the object entity (who it was done to)
                //  identity: the id of the subject entity (who did it)
                //  type: what was done (CREATE, UPDATE, ...)
                //  timestamp: hammertime (milliseconds since unix epoch)
                //  nanotime: ??? (does not appear to be nanoseconds since unix epoch)
                //  body: the full content of the object entity
                this.emit(object);
            };

            //now the socket is opened we can resolve our promise with a client API
            ws.onopen = () => {
                //actually if we send a bad token, we need to wait for a moment before resolving.
                //otherwise, the connection will shut straightaway.
                //testing shows this usually takes about 10ms, but the largest time I saw was
                //32ms. We use 100ms just in case. This makes the initial connection slower, but
                //is a one off cost.
                setTimeout(() => {
                    if (hasResolved) {
                        //crap, we did fail.
                        return;
                    }
                    //now mark this as resolving so we don't invalidate our token accidentally.
                    hasResolved = true;
                    resolve(ws);
                    //now add all our filters.
                    this.filters.forEach(filter => {
                        ws.send(
                            JSON.stringify({
                                type: "register",
                                args: filter
                            })
                        );
                    });
                }, 100);
            };
        });
    }
}
