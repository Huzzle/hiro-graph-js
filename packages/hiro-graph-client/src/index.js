import Token from "./token";
import Connection from "./connection";
import * as Errors from "./errors";

export { Token, Connection, Errors };
