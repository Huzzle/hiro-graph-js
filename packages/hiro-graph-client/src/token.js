/**
 *  Represents a GraphIT Access Token.
 *
 *  Access is always a promise, as we may have to renew it.
 *
 *  both `onInvalidate()` and `getToken()` can return promises.
 *  `getMeta` should return synchronously.
 */
export default class Token {
    constructor({ onInvalidate = () => {}, getMeta = () => {}, getToken }) {
        this._onInvalidate = onInvalidate;
        this._getMeta = getMeta;
        this._getToken = getToken;
    }

    get() {
        if (!this._tokenPromise) {
            this._tokenPromise = Promise.resolve(this._getToken())
                .catch(err => {
                    //flatten this, so another attempt can be made
                    this._tokenPromise = null;
                    //returning a bogus token (null), should hopefully trigger this again,
                    //waiting subscribers will receive it, fail, and try again.
                    console.error("ERROR GETTING TOKEN!", err);
                    return null;
                })
                .then(token => {
                    this._invalidated = false; //this resolved one way or another...
                    return token;
                });
        }
        return this._tokenPromise;
    }

    invalidate() {
        if (this._invalidated) {
            return;
        }
        this._tokenPromise = false;
        this._invalidated = true;
        return Promise.resolve(this._onInvalidate()).catch(() => {});
    }

    //returns nothing by default, but you may want to synchronously return the
    //last valid token, the time is was acquired, it's expiry, a refresh token, etc...
    meta() {
        return this._getMeta();
    }
}
