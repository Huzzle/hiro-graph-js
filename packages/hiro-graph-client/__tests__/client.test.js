/* eslint-env jest */
import createMockClient from "../src/mock";
import Token from "../src/token";
import * as errors from "../src/errors";

// test all know good response types. all known error types.
// edge cases like 404, or 409 being OK responses (connect/disconnect)
// bad token causing failure
// etc...
describe("Client Requests", () => {
    const client = createMockClient();

    const tests = [
        {
            name: "me",
            fn: () => client.me()
        },
        {
            name: "info",
            fn: () => client.info()
        },
        {
            name: "get",
            fn: () => client.get("some-id")
        },
        {
            name: "create",
            fn: () => client.create("some-type", { some: "prop" })
        },
        {
            name: "create (waitForIndex)",
            fn: () =>
                client.create(
                    "some-type",
                    { some: "prop" },
                    { waitForIndex: true }
                )
        },
        {
            name: "update",
            fn: () => client.update("some-id", { "/foo": "bar" })
        },
        {
            name: "update (waitForIndex)",
            fn: () =>
                client.update(
                    "some-id",
                    { "/foo": "bar" },
                    { waitForIndex: true }
                )
        },
        {
            name: "replace",
            fn: () => client.replace("some-id", { "/foo": "bar" })
        },
        {
            name: "replace (createIfNotExists)",
            fn: () =>
                client.replace(
                    "some-id",
                    { "/foo": "bar" },
                    { createIfNotExists: "some-type" }
                )
        },
        {
            name: "replace (waitForIndex)",
            fn: () =>
                client.replace(
                    "some-id",
                    { "/foo": "bar" },
                    { waitForIndex: true }
                )
        },
        {
            name: "delete",
            fn: () => client.delete("some-id")
        },
        {
            name: "delete (waitForIndex)",
            fn: () => client.delete("some-id", { waitForIndex: true })
        },
        {
            name: "ids",
            fn: () => client.ids(["a", "b", "c"])
        },
        {
            name: "connect",
            fn: () => client.connect("some-verb", "the-in-id", "the-out-id")
        },
        {
            name: "disconnect",
            fn: () => client.disconnect("some-verb", "the-in-id", "the-out-id")
        },
        {
            name: "lucene (basic)",
            fn: () => client.lucene("*:*")
        },
        {
            name: "lucene (limits)",
            fn: () => client.lucene("*:*", { limit: -1, offset: 1000 })
        },
        {
            name: "lucene (placeholders)",
            fn: () => client.lucene("/foo:$bar", { bar: "123" })
        },
        {
            name: "lucene (count)",
            fn: () => client.lucene("*:*", { count: true })
        },
        {
            name: "lucene (order)",
            fn: () =>
                client.lucene("*:*", { order: ["/field1 asc", "/field2 desc"] })
        },
        {
            name: "lucene (fields)",
            fn: () => client.lucene("*:*", { fields: ["/field1", "/field2"] })
        },
        {
            name: "gremlin",
            fn: () => client.gremlin("root-id", "pow()")
        }
    ];

    beforeEach(() => {
        client.resetMockTransport();
    });

    tests.forEach(({ name, fn }) => {
        it(`should call ${name} correctly`, async () => {
            client.enqueueMockResponse({});
            // we just have to make sure it resolves period.
            // we will handle expected values for those functions that don't
            // pass the responses right through directly.
            await expect(fn().then(() => true)).resolves.toEqual(true);
            // we only want the three main properties of the request
            const { type, headers, body } = client.retrieveLastRequest();
            expect({ type, headers, body }).toMatchSnapshot();
        });
    });
});

describe("Client Response handling", () => {
    const fn = jest.fn();
    const token = new Token({
        getToken: () => "token",
        onInvalidate: fn
    });
    const client = createMockClient(token);

    beforeEach(async () => {
        client.resetMockTransport();
        await token.get(); // we want to ensure the token invalidation state is reset each time.
        fn.mockClear();
    });

    const retryableErrors = [
        {
            name: "transaction fail (888)",
            err: errors.transactionFail()
        },
        {
            name: "unauthorized (401)",
            err: errors.unauthorized()
        },
        {
            name: "generic retryable error",
            err: Object.assign(new Error(), { isRetryable: true })
        }
    ];

    retryableErrors.forEach(({ name, err }) => {
        it(`should retry once on '${name}'`, async () => {
            client.enqueueMockResponse(err, "ok");
            await expect(client.me()).resolves.toBe("ok");

            client.enqueueMockResponse(err, err);
            await expect(client.me()).rejects.toBe(err);
        });
    });

    it("should unconditionally retry for `connection closed before send`", async () => {
        const err = errors.connectionClosedBeforeSend;
        // many connection closed, one other error in the middle, should still resolve "ok"
        /* prettier-ignore */
        client.enqueueMockResponse(err, err, err, err, errors.transactionFail(), err, err, err, err, "ok");
        await expect(client.me()).resolves.toBe("ok");
    });

    it("should handle `conflict` as OK for connect", async () => {
        const conflict = errors.conflict();
        const forbidden = errors.forbidden();
        client.enqueueMockResponse(conflict, forbidden);
        await expect(
            client.connect("foo", "bar", "baz")
        ).resolves.toBeUndefined(); // connect returns nothing on success
        await expect(client.connect("foo", "bar", "baz")).rejects.toBe(
            forbidden
        );
    });

    it("should handle `conflict` and `not found` as OK for disconnect", async () => {
        const conflict = errors.conflict();
        const notFound = errors.notFound();
        const forbidden = errors.forbidden();
        client.enqueueMockResponse(conflict, forbidden, notFound);
        await expect(
            client.disconnect("foo", "bar", "baz")
        ).resolves.toBeUndefined(); // disconnect returns nothing on success

        await expect(client.disconnect("foo", "bar", "baz")).rejects.toBe(
            forbidden
        );
        await expect(
            client.disconnect("foo", "bar", "baz")
        ).resolves.toBeUndefined(); // disconnect returns nothing on success
    });

    const dedupableMethods = [
        ["me"],
        ["get", "some-id"],
        ["lucene", "*:*"],
        ["ids", ["a", "b"]],
        ["gremlin", "root-id", "outE()"]
    ];

    dedupableMethods.forEach(([method, ...args]) => {
        it(
            "should correctly dedup the requests for method: " + method,
            async () => {
                let resolve;
                const promise = new Promise(r => (resolve = r));
                // note that we only enqueue a single result
                // so both client methods MUST only make a single
                // request or it will fail Bad Request
                client.enqueueMockResponse(promise);
                const req1 = client[method](...args);
                const req2 = client[method](...args);
                resolve({ foo: "bar" });
                const res1 = await req1;
                const res2 = await req2;

                // these expectations is that bot res1 and res2 returned the same result
                // but not the same object.
                expect(res1).toEqual(res2);
                expect(res1).not.toBe(res2);
            }
        );
    });
    it("should invalidate the token on `unauthorized`", async () => {
        // ok first, then non invalidation error, then invalidation error, then notfound.
        // we need another error after, so the retry can pick it up
        const notFound = errors.notFound();
        const unauthorized = errors.unauthorized();
        client.enqueueMockResponse(
            "OK",
            notFound,
            unauthorized,
            notFound,
            "OK"
        );

        await expect(client.me()).resolves.toBe("OK");
        expect(fn).not.toHaveBeenCalled();

        await expect(client.me()).rejects.toBe(notFound);
        expect(fn).not.toHaveBeenCalled();

        //this one should invalidate
        await expect(client.me()).rejects.toBe(notFound);
        expect(fn).toHaveBeenCalledTimes(1);

        await expect(client.me()).resolves.toBe("OK");
        expect(fn).toHaveBeenCalledTimes(1);
    });
});
