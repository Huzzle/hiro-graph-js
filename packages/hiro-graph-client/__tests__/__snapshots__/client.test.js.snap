// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client Requests should call connect correctly 1`] = `
Object {
  "body": Object {
    "in": "the-in-id",
    "out": "the-out-id",
  },
  "headers": Object {
    "ogit/_type": "some-verb",
  },
  "type": "connect",
}
`;

exports[`Client Requests should call create (waitForIndex) correctly 1`] = `
Object {
  "body": Object {
    "some": "prop",
  },
  "headers": Object {
    "ogit/_type": "some-type",
    "waitForIndex": "true",
  },
  "type": "create",
}
`;

exports[`Client Requests should call create correctly 1`] = `
Object {
  "body": Object {
    "some": "prop",
  },
  "headers": Object {
    "ogit/_type": "some-type",
  },
  "type": "create",
}
`;

exports[`Client Requests should call delete (waitForIndex) correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "ogit/_id": "some-id",
    "waitForIndex": "true",
  },
  "type": "delete",
}
`;

exports[`Client Requests should call delete correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "ogit/_id": "some-id",
  },
  "type": "delete",
}
`;

exports[`Client Requests should call disconnect correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "ogit/_id": "the-out-id$$some-verb$$the-in-id",
  },
  "type": "delete",
}
`;

exports[`Client Requests should call get correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "ogit/_id": "some-id",
  },
  "type": "get",
}
`;

exports[`Client Requests should call gremlin correctly 1`] = `
Object {
  "body": Object {
    "query": "pow()",
    "root": "root-id",
  },
  "headers": Object {
    "type": "gremlin",
  },
  "type": "query",
}
`;

exports[`Client Requests should call ids correctly 1`] = `
Object {
  "body": Object {
    "query": "a,b,c",
  },
  "headers": Object {
    "type": "ids",
  },
  "type": "query",
}
`;

exports[`Client Requests should call info correctly 1`] = `
Object {
  "body": undefined,
  "headers": undefined,
  "type": "info",
}
`;

exports[`Client Requests should call lucene (basic) correctly 1`] = `
Object {
  "body": Object {
    "count": false,
    "limit": 50,
    "offset": 0,
    "query": "*:*",
  },
  "headers": Object {
    "type": "vertices",
  },
  "type": "query",
}
`;

exports[`Client Requests should call lucene (count) correctly 1`] = `
Object {
  "body": Object {
    "count": true,
    "limit": 50,
    "offset": 0,
    "query": "*:*",
  },
  "headers": Object {
    "type": "vertices",
  },
  "type": "query",
}
`;

exports[`Client Requests should call lucene (fields) correctly 1`] = `
Object {
  "body": Object {
    "count": false,
    "fields": "/field1,/field2",
    "limit": 50,
    "offset": 0,
    "query": "*:*",
  },
  "headers": Object {
    "type": "vertices",
  },
  "type": "query",
}
`;

exports[`Client Requests should call lucene (limits) correctly 1`] = `
Object {
  "body": Object {
    "count": false,
    "limit": -1,
    "offset": 1000,
    "query": "*:*",
  },
  "headers": Object {
    "type": "vertices",
  },
  "type": "query",
}
`;

exports[`Client Requests should call lucene (order) correctly 1`] = `
Object {
  "body": Object {
    "count": false,
    "limit": 50,
    "offset": 0,
    "order": "/field1 asc,/field2 desc",
    "query": "*:*",
  },
  "headers": Object {
    "type": "vertices",
  },
  "type": "query",
}
`;

exports[`Client Requests should call lucene (placeholders) correctly 1`] = `
Object {
  "body": Object {
    "bar": "123",
    "count": false,
    "limit": 50,
    "offset": 0,
    "query": "/foo:$bar",
  },
  "headers": Object {
    "type": "vertices",
  },
  "type": "query",
}
`;

exports[`Client Requests should call me correctly 1`] = `
Object {
  "body": Object {},
  "headers": Object {},
  "type": "me",
}
`;

exports[`Client Requests should call replace (createIfNotExists) correctly 1`] = `
Object {
  "body": Object {
    "/foo": "bar",
  },
  "headers": Object {
    "createIfNotExists": "true",
    "ogit/_id": "some-id",
    "ogit/_type": "some-type",
  },
  "type": "replace",
}
`;

exports[`Client Requests should call replace (waitForIndex) correctly 1`] = `
Object {
  "body": Object {
    "/foo": "bar",
  },
  "headers": Object {
    "ogit/_id": "some-id",
    "waitForIndex": "true",
  },
  "type": "replace",
}
`;

exports[`Client Requests should call replace correctly 1`] = `
Object {
  "body": Object {
    "/foo": "bar",
  },
  "headers": Object {
    "ogit/_id": "some-id",
  },
  "type": "replace",
}
`;

exports[`Client Requests should call update (waitForIndex) correctly 1`] = `
Object {
  "body": Object {
    "/foo": "bar",
  },
  "headers": Object {
    "ogit/_id": "some-id",
    "waitForIndex": "true",
  },
  "type": "update",
}
`;

exports[`Client Requests should call update correctly 1`] = `
Object {
  "body": Object {
    "/foo": "bar",
  },
  "headers": Object {
    "ogit/_id": "some-id",
  },
  "type": "update",
}
`;
